### PR TITLE
[Bugfix][Model] Make PersonaPlex Code2Wav input delta-only

### PR DESCRIPTION
**Pull request by jeffaa729**
## Purpose

Addresses item 4 of the PersonaPlex Stage 1 delta-only Code2Wav task in #7389.

The async connector already sends delta codec frames, but `PersonaPlexCode2Wav` retained the complete CPU history and performed a content-based prefix comparison or concatenation for every chunk.

The prefix comparison was also incorrect for valid repeated codec values. If a new chunk begins with the same codes as the previous chunk, the old implementation can mistake it for a cumulative payload and discard it as an empty suffix. 

This PR removes `_new_code_suffix` and `_request_codes`. In async-chunk mode, every connector delta is passed directly to `_decode_streaming_frames`.

The registered full-payload mode is preserved. When `async_chunk=False`, a per-request flag ensures that the cumulative payload is consumed only once when chunked prefill causes it to appear in multiple Stage 1 forwards. The flag is cleared by `on_requests_finished()`.

The regression coverage verifies that:

- 1,000 one-frame deltas produce exactly 1,000 decoder calls, each with shape `(2, 1)`;
- two consecutive delta frames containing identical codec values are both decoded;
- a full payload is consumed only once across repeated forwards; and
- the full payload can be consumed again after request cleanup.

### Benchmark

A timing assertion was not committed because host-side microbenchmarks are noisy and can make CI flaky. The following non-gating before-and-after measurement evaluates the async-chunk Code2Wav input path.

The benchmark first ran complete sessions for 50 ms of host-thread CPU time as warmup. It then measured 100 independent sessions, each containing 1,000 one-frame delta chunks. Each table row reports the mean per-chunk host-thread CPU time for the corresponding 100-frame interval across those sessions.

Environment: AMD Ryzen 9 8945HS, WSL2, Python 3.12.3, vLLM `0.29.0`.

| Frames           | Before (ns/chunk) | After (ns/chunk) |
| :--------------- | ----------------: | ---------------: |
| 0000-0100        |            34,635 |           22,166 |
| 0100-0200        |            34,392 |           22,750 |
| 0200-0300        |            34,628 |           23,870 |
| 0300-0400        |            34,418 |           23,487 |
| 0400-0500        |            33,310 |           22,852 |
| 0500-0600        |            33,300 |           22,737 |
| 0600-0700        |            34,641 |           22,922 |
| 0700-0800        |            33,675 |           23,345 |
| 0800-0900        |            34,114 |           22,351 |
| 0900-1000        |            33,402 |           23,605 |
| **Overall mean** |        **34,052** |       **23,009** |

The overall mean host-thread CPU time decreased by approximately 32.4%.

## Test Plan

**vLLM Version:** `0.29.0`

```bash
pytest -s -v \
  tests/model_executor/models/personaplex/test_streaming_code2wav.py

pre-commit run --files \
  vllm_omni/model_executor/models/personaplex/personaplex_code2wav.py \
  tests/model_executor/models/personaplex/test_streaming_code2wav.py
```

## Test Result

The PersonaPlex Code2Wav test file passed with `11 passed, 14 warnings in 1.11s`. All applicable pre-commit hooks passed, including Ruff, mypy, CI marks, SPDX, forbidden-import, `torch.cuda`, and DCO checks.

## AI Assistance

AI assistance: Used Codex to analyze issue #7389, develop and review the implementation and regression tests, and draft this PR description. I reviewed the changes and ran the reported validation locally.

### CODE REVIEW

**Comment by jeffaa729** [comment]
@vllm-omni-review-bot please review this PR.

**Comment by jeffaa729** [comment]
Self-review:

I checked that:

- `forward()` still validates codebook divisibility, preserves the codebook-major `(k, F)` shape and device transfer, and handles profile and malformed inputs unchanged;
- request-ID fallback, codec-slot isolation, request cleanup, and slot reuse remain covered by tests;
- the regression exercises `forward()` with 1,000 one-frame deltas and verifies 1,000 `(2, 1)` decoder inputs with no CPU history concatenation or prefix comparison;
- this PR does not change Stage 1 PCM `output_kind` or the data-plane cursor, so #7389 item 6 remains separate;
- the diff is limited to the PersonaPlex Code2Wav implementation and its CPU tests, with no user-facing API, configuration, or documentation changes.

Validation: `9 passed, 14 warnings in 1.40s`; all applicable pre-commit hooks and `git diff --check` passed.

AI assistance: Used Codex for the testing, benchmarks, and this self-review.

**Review by linyueqian — CHANGES_REQUESTED** [review]
Reviewed at head `e6952dd27`. The async path works well and fixes a latent bug on `main`, but the full-payload path regresses, so I am requesting one change before merge.

What works:
On the async path, the payload is a true delta. The producer emits `emit_frames` de-delayed frames from raw rows `[:emit_frames + 1]` and keeps the successor row only as producer state (`stage_input_processors/personaplex.py:229-236`), so consecutive payloads share no acoustic frame. The consumer commit replaces `prompt_token_ids` per chunk and skips 1-D `codes.audio` from `additional_information` (`chunk_transfer_adapter.py:611-635`), so `runtime_info` never carries cumulative codes there. End to end on one H100 80GB, a stack with this PR ran the two-session lifecycle driver with statistics identical to `main`, and the output WAVs hashed byte-identical across the baseline, a second baseline run, and the stack, so the async path is proven unchanged.

A detail worth adding to the description: `main`'s prefix compare (`personaplex_code2wav.py:273-278` on `main`) can spuriously match when a chunk's leading codes equal the previous chunk. Constant silence codes make this likely with a 1-frame initial chunk, causing it to silently drop an 80 ms frame. The head fixes this bug, which justifies the Bugfix tag more than host time does.

The regression is at `personaplex_code2wav.py:57-58` and `:210-224` at head. The full-payload path is a registered mode, not a leftover. `pipeline.py:53` registers `talker2code2wav_full_payload` and `:67-71` registers the sync placeholder processor, while `stage_config.py:887-893` selects them whenever `async_chunk` is false. That producer sends the whole cumulative code payload once per finished talker request. On the consumer, that payload persists in the runner's intermediate buffer and returns on every forward for the request (`gpu_model_runner.py:1361-1364`, `:1401-1402`). One delivery does not mean one forward. With `enable_chunked_prefill` and a placeholder longer than `max_num_batched_tokens` (placeholders are eight tokens per frame, so 128 frames is 1,024 tokens against a 512 budget), Stage 1 runs the request in two forwards, and head decodes and emits the entire payload on both while advancing the streaming decoder state. The guard on `main` suppressed that second consumption. The PR also removed the only test that covered cumulative input.

Fix direction: consume a full payload exactly once. You could use a per-request consumed flag set on the first forward and cleared in `on_requests_finished`, or drop the full-payload registration and document `async_chunk: true` as required for PersonaPlex. Either way, restore a regression test for the full-payload path. Please do not bring back content-based prefix guessing.

Non-blocking notes: `test_delta_codes_skip_cpu_history` spies global `torch.cat` and `torch.equal` and asserts zero calls, which pins implementation rather than contract. The decode call count and shapes are the assertions that matter. Also, #7453 instruments `delta_kf`, which this PR removes, so please coordinate with shihongzhi on whichever lands second.

**Comment by jeffaa729** [comment]
@linyueqian I addressed the requested full-payload change.

In `personaplex_code2wav.py`, `PersonaPlexCode2Wav` now reads the resolved async_chunk setting from model_config. For async_chunk=False, it records when a request’s full payload has been consumed, skips the same payload on subsequent chunk-prefill forwards, and clears the record in on_requests_finished(). The async path remains unchanged and decodes every incoming delta.

Two regression tests are added in `test_streaming_code2wav.py`:

`test_identical_consecutive_delta_frames_are_both_decoded` verifies that two consecutive deltas containing identical codec values are both decoded, preventing the old content-based prefix guessing from dropping a valid frame.

`test_full_payload_is_consumed_once_across_forwards` verifies that a full payload is decoded on its first forward, skipped when presented again during chunked prefill, and accepted again after request cleanup.

I also updated `test_delta_codes_skip_cpu_history` in the same test file to remove the global torch.cat and torch.equal spies.

Note for #7453 and @shihongzhi: this PR removes `_new_code_suffix`, so the local `delta_kf` variable no longer exists; `codes_kf` is passed directly to `_decode_streaming_frames `and represents the incoming delta when `async_chunk=True`.

**Review by linyueqian — APPROVED** [review]
Re-reviewed at head `2fd1dcc17`. The change request is met exactly as asked, so approving.

The full payload is now consumed once per request id: `_async_chunk` is read from the per-stage model config (`personaplex_code2wav.py:81`; the field exists on `OmniModelConfig` and `stage_config.py:969` writes it into every stage), the flag is set on the first forward that reaches decode with a well-formed payload and skipped afterwards (`:226-229`), and `on_requests_finished` discards it (`:321-324`), which also covers aborts since they land in `finished_req_ids`. The async path never touches the flag, and `torch.equal`, `_new_code_suffix` and `_request_codes` are gone, so no content-based guessing came back.

Tests drive the production `forward`: `test_full_payload_is_consumed_once_across_forwards` sends the same payload on two consecutive forwards in sync mode and asserts 8 samples then 0 with no extra decode calls, then clears via `on_requests_finished` and decodes again; `test_identical_consecutive_delta_frames_are_both_decoded` pins the frame drop the old prefix compare had; the 1000-delta test no longer spies global `torch.cat`. The PR's own diff is still just the model file and its test, merges cleanly onto current `main`, and `git diff --check` is clean.

Two optional notes. `getattr(..., "async_chunk", False)` would silently take the full-payload branch if the attribute were ever missing; it is not missing anywhere today, but reading the attribute directly would fail loudly instead. And in the unshipped combination of duplex sessions with `async_chunk: false`, the flag persists across a resumable segment for the same request id; a one-line docstring note is enough.
